File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
#include <cmath>

using namespace std;

void Div(int n, map<int, int>& Prime)
{
    if(n == 1) return;
    
    vector<int> temp;
    while(n > 1)
    {
        bool canDiv = false;
        for(int i = 2; i <= sqrt(n); i++)
        {
            if(n % i == 0)
            {
                temp.push_back(i);
                n /= i;
                
                canDiv = true;
                break;
            }
        }

        if(!canDiv)
        {
            temp.push_back(n);
            break;
        }
    }
    
    map<int,int> Current;
    for(auto num : temp)
    {
        Current[num]++;
    }
    
    for(auto [num, cnt] : Current)
    {
        if(Prime[num] < cnt)
        {
            Prime[num] += (cnt - Prime[num]);
        }
    }
}

int solution(vector<int> arr) {
    int answer = 1;
    
    map<int, int> Prime;
    
    for(int num : arr)
    {
        Div(num, Prime);
    }
    
    for(auto [num, cnt] : Prime)
    {
        // cout << num << ": " << cnt << ", ";
        answer *= pow(num,cnt);
        
    }
    
    return answer;
}

File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.cs
using System;
using System.Linq;

public class Solution {
    
    int GCD(int a, int b)
    {
        if(a < b) (b, a) = (a, b);
        
        if(b == 0) return a;
        return GCD(b, a%b);
    }
    
    public int solution(int[] arr) {
        
        int gcd = GCD(arr[0], arr[1]);
        int num = arr[0] / gcd * arr[1];
        
        for(int i = 1; i < arr.Length; i++)
        {
            gcd = GCD(num, arr[i]);
            num = num * arr[i] / gcd;
        }

        return num;
    }
}